### ISSUE TITLE
[R] Is it possible to train Mxnet symbols and a GLM pretrained model as loss?

### ISSUE THREAD

**Issue by some-guy1** [apache/incubator-mxnet#10890]
The idea is very similar to a GAN, but instead of two neural networks I want to use a neural network generator and a prebuilt GLM as the discriminator.

Is this possible using the Mxnet symbols?

Example pseudo-code:
```
data <- mx.symbol.Variable('data')
label <- mx.symbol.Variable('label')
dconv <- mx.symbol.Deconvolution(data= data , kernel = c(3,3), stride=c(3,3), num_filter = 5)
norm = mx.symbol.BatchNorm(data= dconv, fix_gamma=FALSE)
act1 <- mx.symbol.LeakyReLU(data= norm, act_type = "leaky", name="act_3")

#GLM that was already built and will not change
class = abs(predict.glm(act1) - label)
NN_Model <- mx.symbol.MakeLoss(class )
```

**Comment by harryprince** [comment]
pseudo-code seems not works, because predict.glm is implemented in R instead of mxnet. the parameter flat is just a symbol.

**Comment by jeremiedb** [comment]
Since a GLM predictions can be expressed as a matrix product and some link function transformation, it should be relatively straightforward to convert into a symbolic representation. 
I think you could use the FullyCOnnected operator with num_hidden = 1 and specify the GLM weights as fixed parameters in model training.

**Comment by some-guy1** [comment]
Ok I want to be able to use any other technique.  GLM, forests, SVR, boosting, etc etc.

After looking through some of the source code I noticed that the values in the custom metric are no longer symbols, but are actual values.  So, I tried to first turn off the loss by:

```
last_activation <- mx.symbol.Activation(data= conv_6, act_type = "sigmoid", name="Last_act_8")
last <- mx.symbol.concat(data=c(last_activation,data), num.args = 2, dim = 1, name='cato_last')
NN_Model <- mx.symbol.MakeLoss(last )
```
Then I created a custom metric:

```
mx.metric.GAN <- mx.metric.custom("GAN", function(label, pred)
{
   res = 0 
   for(i in 1:batch_size)
   {      
     label.data = tail(pred[,i], 9)
     pred.data = head(pred[,i], -9)     
     dim(pred.data) <- c(238,238,3,1)     
     flat = nn.function(pred.data[,,,1])
     svr.pred = predict(svm.model, t(flat))
     res = res + (svr.pred - label.data[1] )^2       
   }   
   return(as.array(res))
})
```

This seems to actually call the other R functions that I want.  But, now mxnet appears to be attempting to maximize the metric instead of minimize it.

Seems like I am very close to getting this to work, but I am not sure if the way I have manipulated the custom loss function is appropriate.

**Comment by harryprince** [comment]
@some-guy1 
it seems this is the answer:
https://www.r-bloggers.com/deep-learning-with-mxnetr/

```
# Define the input data
data <- mx.symbol.Variable("data")
# A fully connected hidden layer
# data: input source
# num_hidden: number of neurons in this layer
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)

# Use linear regression for the output layer
lro <- mx.symbol.LinearRegressionOutput(fc1)
```
